### PR TITLE
Coxeter

### PR DESCRIPTION
**Pull request by skachano**
Because of issues in the last pull request (specifically concerning Strong_collapse and Toplex_map that did not compile), I decided to make a fresh copy of the master from Gudhi and copy the Coxeter_triangulation directory there.

### CODE REVIEW

File: src/Coxeter_triangulation/doc/intro_coxeter_triangulation.h [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+\li \ref Gudhi::coxeter_triangulation::Cartesian_product "Cartesian_product(functions...)" expresses the Cartesian product \f$F_1^{-1}(0) \times \ldots \times F_k^{-1}(0)\f$ of multiple implicit manifolds as an implicit manifold.
+  For convenience, a static function \ref Gudhi::coxeter_triangulation::make_product_function() "make_product_function(functions...)" is provided that takes a pack of function-typed objects as the argument.
+\li \ref Gudhi::coxeter_triangulation::Embed_in_Rd "Embed_in_Rd(F, d)" expresses an implicit manifold given as the zero-set of a function \f$F\f$ embedded in a higher-dimensional Euclidean space \f$\mathbb{R}^d\f$.
+  For convenience, a static function \ref Gudhi::coxeter_triangulation::make_embedding() "make_embedding(F, d)" is provided.
+\li \ref Gudhi::coxeter_triangulation::Linear_transformation "Linear_transformation(F, M)" applies a linear transformation given by a matrix \f$M\f$ on an implicit manifold given as the zero-set of the function \f$F\f$.
+  For convenience, a static function \ref Gudhi::coxeter_triangulation::linear_transformation() "linear_transformation(F, M)" is provided.
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/doc/intro_coxeter_triangulation.h]:
```suggestion
  For convenience, a static function \ref Gudhi::coxeter_triangulation::make_linear_transformation "linear_transformation(F, M)" is provided.
```

File: src/Coxeter_triangulation/example/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 2.6)
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/example/CMakeLists.txt]: You can remove this line as it is now done at top level

File: src/Coxeter_triangulation/example/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+project(Coxeter_triangulation_example)
+
+add_executable ( Coxeter_triangulation_manifold_tracing_flat_torus_with_boundary_example manifold_tracing_flat_torus_with_boundary.cpp )
+add_executable ( Coxeter_triangulation_manifold_tracing_custom_function_example manifold_tracing_custom_function.cpp )
+
+if (TBB_FOUND)
+   target_link_libraries(Coxeter_triangulation_manifold_tracing_flat_torus_with_boundary_example ${TBB_LIBRARIES})
+   target_link_libraries(Coxeter_triangulation_manifold_tracing_custom_function_example ${TBB_LIBRARIES})
+endif()
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/example/CMakeLists.txt]: You could, at least test your examples:
```suggestion
endif()
add_test(NAME Coxeter_triangulation_manifold_tracing_flat_torus_with_boundary_example COMMAND $<TARGET_FILE:Coxeter_triangulation_manifold_tracing_flat_torus_with_boundary_example>)
add_test(NAME Coxeter_triangulation_manifold_tracing_custom_function_example COMMAND $<TARGET_FILE:Coxeter_triangulation_manifold_tracing_custom_function_examplele>)
```

**Comment by mglisse** [comment]
> src/Coxeter_triangulation/include/gudhi/Functions/random_orthogonal_matrix.h makes your module depend on CGAL (which means GPL license) only because of `CGAL::Random_points_on_sphere_d`.

CGAL package Generator seems to be LGPL.

> This function is also available in src/common/include/gudhi/random_point_generators.h thanks to `generate_points_on_sphere_d`. It would be better to use this function.

generate_points_on_sphere_d uses CGAL::Random_points_on_sphere_d, so that wouldn't remove the dependency.

**Comment by VincentRouvreau** [comment]
> CGAL package Generator seems to be LGPL.
> generate_points_on_sphere_d uses CGAL::Random_points_on_sphere_d, so that wouldn't remove the dependency.

Oh yes, you are completely right, just forget my point...

**Comment by VincentRouvreau** [comment]
All tests cover really nicely the code, except `elementary_increment_optim_3()` function. Is it dead code or not ?
`include/gudhi/IO/*` is only tested in examples but I am ok to keep it.

File: src/Coxeter_triangulation/include/gudhi/Coxeter_triangulation.h [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	  = Permutahedral_representation<std::vector<int>, std::vector<std::vector<std::size_t> > > >
+class Coxeter_triangulation : public Freudenthal_triangulation<Permutahedral_representation_> {
+
+  typedef Eigen::MatrixXd Matrix;
+
+  Matrix root_matrix(unsigned d) {
+    Matrix cartan(d,d);
+    for (unsigned i = 0; i < d; i++) {
+      cartan(i,i) = 1.0;
+    }
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Coxeter_triangulation.h]: Didn't test it but you could use `Ones`:
```suggestion
Matrix cartan(d,d) = Matrix::Ones(d,d);
```

File: src/Coxeter_triangulation/include/gudhi/Coxeter_triangulation.h [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    for (unsigned i = 0; i < d; ++i)
+      sqrt_diag(i) = std::sqrt(saes.eigenvalues()[i]);
+
+    Matrix lower(d,d);
+    for (unsigned i = 0; i < d; i++)
+      for (unsigned j = 0; j < d; j++)
+	if (i < j)
+	  lower(i,j) = 0;
+	else
+	  lower(i,j) = 1;
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Coxeter_triangulation.h]: Here you could use `Ones` and [Eigen triangular view](https://eigen.tuxfamily.org/dox/classEigen_1_1MatrixBase.html#a56665aa894f49f2765291fee0eaeb9c6)

File: src/Coxeter_triangulation/include/gudhi/Coxeter_triangulation.h [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+#ifndef COXETER_TRIANGULATION_H_
+#define COXETER_TRIANGULATION_H_
+
+#include <stack>
+#include <map>
+#include <numeric> //iota
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Coxeter_triangulation.h]: These include are not used but you missed this one:
```suggestion
#include <vector>
```

File: src/Coxeter_triangulation/include/gudhi/Cell_complex.h [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+ */
+
+#ifndef CELL_COMPLEX_H_
+#define CELL_COMPLEX_H_
+
+#include <Eigen/Dense> 
+
+#include <algorithm>
+#include <gudhi/Permutahedral_representation/Simplex_comparator.h>
+#include <fstream> // for Hasse_diagram_persistence.h
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Cell_complex.h]: Include what you use:
```suggestion
#include <vector>
#include <map>
#include <utility>  // for std::make_pair

#include <gudhi/Permutahedral_representation/Simplex_comparator.h>
```

If `#include <fstream>` is required by Hasse_diagram_persistence.h, it is up to this file to include it.

File: src/Coxeter_triangulation/include/gudhi/Freudenthal_triangulation.h [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+ *    Modification(s):
+ *      - YYYY/MM Author: Description of the modification
+ */
+
+#ifndef FREUDENTHAL_TRIANGULATION_H_
+#define FREUDENTHAL_TRIANGULATION_H_
+
+#include <stack>
+#include <map>
+#include <numeric> //iota
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Freudenthal_triangulation.h]: Include what you use:
```suggestion
#include <vector>
#include <algorithm>  // for std::sort
#include <cmath> // for std::floor
#include <numeric> // for std::iota
```

File: src/Coxeter_triangulation/include/gudhi/Freudenthal_triangulation.h [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+   * triangulation that shares the origin, but is scaled by a given factor.
+   * This parameter can be useful to simulate the computation of Cartesian coordinates 
+   * of the barycenter of a simplex in a subdivided triangulation.
+   * @param[in] simplex The query simplex.
+   * @param[in] scale The scale of the triangulation. 
+   */
+  Eigen::VectorXd barycenter(const Simplex_handle& simplex, double scale = 1) const {
+    Eigen::VectorXd res_vector(dimension_);
+    for (size_t i = 0; i < dimension_; ++i)
+      res_vector(i) = 0;
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Freudenthal_triangulation.h]:
```suggestion
res_vector.setZero(dimension_, 1);
```

File: src/Coxeter_triangulation/include/gudhi/Implicit_manifold_intersection_oracle.h [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+  /* Computes the affine coordinates of the intersection point of the implicit manifold
+   * and the affine hull of the simplex. */
+  template <class Simplex_handle, 
+	    class Triangulation>
+  Eigen::VectorXd compute_lambda(const Simplex_handle& simplex,
+				 const Triangulation& triangulation) const {
+    std::size_t cod_d = this->cod_d();
+    Eigen::MatrixXd matrix(cod_d + 1, cod_d + 1);
+    for (std::size_t i = 0; i < cod_d + 1; ++i)
+      matrix(0, i) = 1;
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Implicit_manifold_intersection_oracle.h]:
```suggestion
    matrix.setOnes(cod_d + 1, cod_d + 1);
```

File: src/Coxeter_triangulation/include/gudhi/Permutahedral_representation.h [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+ */
+
+#ifndef PERMUTAHEDRAL_REPRESENTATION_H_
+#define PERMUTAHEDRAL_REPRESENTATION_H_
+
+#include <gudhi/Permutahedral_representation/Permutahedral_representation_iterators.h>
+
+#include <unordered_set>
+#include <iostream>
+
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Permutahedral_representation.h]:
```suggestion
#include <utility>  // for std::make_pair
```

File: src/Coxeter_triangulation/include/gudhi/Cell_complex/Hasse_diagram_cell.h [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+ */
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <string>
+#include <sstream>
+#include <type_traits>
+#include <cstdlib>
+#include <cstdio>
```
Comment (VincentRouvreau) [inline src/Coxeter_triangulation/include/gudhi/Cell_complex/Hasse_diagram_cell.h]: Include what you use:
```suggestion
#include <vector>
#include <utility>  // for std::pair
#include <ostream>
#include <string>
#include <type_traits>  // for std::is_same
```

**Comment by VincentRouvreau** [comment]
Integration in progress on https://github.com/VincentRouvreau/gudhi-devel/tree/coxeter_integration

**Pull request closed without merging**